little fixes, extensions to visualizer, ddp, deterministic training

wendywwang opened the pull request:
This PR is a collection of the following changes:

#### 1. Fixes 
1. for training
    - specified data type `int` for `gpu-ids` argument to avoid complaints from pytorch/cuda
    - moved `tensor2im` and `tensor_to_pil` (which calls `tensor2im`) from `deepliif/util/__init__.py` to `deepliif/util/util.py` so they could be referred to properly
2. for inference
    - (not a fix) added `[-2:]` for `cv2.findContours()` to be compatible with newer cv2 version
    - defaulted `eager_mode=True` in `init_nets()` for inference to be able to locate the correct model file names
   
#### 2. Extensions to Visualizer
- implemented an option `remote` to isolate model training and visualization
  - it saves the input data as pickled snapshots files, similar to what tensorboard would expect snapshots files; you will run the file `run_local_visualizer.py` (provide a directory to the snapshots) to launch the visualizer and display the changes 
  - this is useful when 1) training and visualization cannot happen in the same environment (this itself can be caused by for example running experiments on a few remote virtual machines while all training progress are preferred to be displayed in one visdom visualizer, each in a different display env), and 2) if we want to preserve a history of snapshots so that we can restore the graphs from an experiment in the past (currently the snapshots overwrites the old ones, but can easily be modified to support a history of snapshots) 
- (not only for visualizer) implemented an option `remote_transfer_cmd` to consume custom module/script that describes how to transfer the snapshots of visdom's input data AND model file
  - an example is the following code snippet (let's call it `mysave.py`), to use which you will pass `--remote-transfer-cmd mysave.save_to_s3` in training:
```
import s3fs

credentials = <s3 credentials>

def save_to_s3(source_path):
  <code to put the file at source_path to s3>

```
  - this is a flexible method to support any custom transferal, and instead of s3 you could use a database, a volume, or anything as long as you know how to put local files to that place through python

#### 3. Distributed Data Parallel 
As recommended by [PyTorch's documentation](https://pytorch.org/docs/stable/notes/cuda.html#use-nn-parallel-distributeddataparallel-instead-of-multiprocessing-or-nn-dataparallel), Distributed Data Parallel (DDP) is a preferred method and is expected to have better scalability as compared to Data Parallel (DP) which has been implemented already, even on a single machine with multiple GPUs. In this PR, DDP is implemented as an option because of backward comparability, and also for more tests to be done before we can confidently remove DP.

The following variables are examined to determine whether the training uses DDP or DP:
- argument `local-rank`: this is an placeholder argument in training, for which a value will be supplied if the user launched DDP using `torchrun` ([doc](https://pytorch.org/docs/stable/elastic/run.html)) or its old version `torch.distributed.launch`. Currently having a valid local rank is viewed as running DDP on a local machine via `torchrun`, and will trigger relevant settings.
- env var `RANK`: this is a variable typically present when running DDP not using `torchrun`; when this env var is available, user is assumed to have spawned the processes in their preferred way already without a need to initialize process group on our side, and the processes could be on multiple nodes (since torchrun itself only deals with the local machine). 

Changes related to DDP mainly are:
- initialize process group in `cli.py`
- apply DDP wrapper class in `deepliif/models/networks.py`
- configure `DistributedDataSampler` in `deepliif/data/__init__.py` and enable shuffle in `cli.py`

Implemented but not enabled features:
- convert batch norm layers to snyc batch norm when using DDP (leads to significantly longer training time)
- save models from only 1 process rather than all processes (hangs at the next iteration)
  - note that saving from all processes vs. 1 doesn't make a difference if we simply want to have the models persisted, because all processes will end up with exactly the same parameters after gradient sync

#### 4. Deterministic Training
A new option `seed` is added to training. Passing a number to this argument will trigger:
- a bunch of settings to control randomness in `cli.py`, such as seeding the random generators and disabling cudnn benchmark (these will slow down the training)
- usage of `ZeroPad2d` in replacement of `ReflectionPad2d` in ResNet Generators, because `ReflectionPad2d` has no deterministic implementation (this could make the training result worse)

Note that this is an experiment feature. If `seed` is used in training, env var `DEEPLIIF_SEED` needs to be set up (any number would be fine) for inference on trained models, because the padding layers are determined based on this env var.

Parmida93 commented:
Please follow these instructions before merging:
1. Add the DDP version of DeepLIIF in a separate directory.
2. Add commands and instructions on how to train and test the model using DDP in its README file.
3. Use the main DeepLIIF model in the deepliff library to avoid inconsistencies in future versions of DeepLIIF (the new base model can be used, just make sure to use the DeepLIIF_model in the deepliif library).

wendywwang commented:
@Parmida93 With multi-gpu training, actions based on the number of total iterations (`total_iters`) might not function as expected 🤔.

For example, for displaying images on visdom, the default value of `--display-freq` is 400. If the batch size we use is 3, and every time this batch size 3 gets added to total_iters, the display image action will not be called until the 1200 iterations (the least common multiple).

Another thing I'm thinking is whether to scale the total iters variable here in the for loop in `train()`. If we use multiple processes, the effective total iterations will be the current iteration number in one process multiplied by the number of processes.

cc @rich-nieto-ibm @drewmibm

ricdodds commented on deepliif/data/__init__.py:
```diff
@@ -74,11 +77,33 @@ def __init__(self, dataset, batch_size, serial_batches, num_threads, max_dataset
+            generator=g
+        )
+
+        self.sampler=sampler            
```
This assignment seems to have no effect and could be removed?

wendywwang replied:
@ricdodds this is needed because in `cli.py` we would call dataloader.sampler.set_epoch() in each epoch, if DDP is used & shuffle is preferred

ricdodds commented on deepliif/data/__init__.py:
```diff
@@ -74,11 +77,33 @@ def __init__(self, dataset, batch_size, serial_batches, num_threads, max_dataset
+        g = torch.Generator()
+        g.manual_seed(0)
+
         self.dataloader = torch.utils.data.DataLoader(
```
Maybe to ensure we're not adding unintended effects in the DataLoader, g (generator) and sampler should be None if this is ran in the traditional setting. Does it make sense?

wendywwang replied:
@ricdodds yes you are right, let me put a conditional check for DEEPLIIF_SEED here to determine whether we use this g when we define self.dataloader (same for the seed_worker function)

wendywwang replied:
fixed

ricdodds commented on deepliif/postprocessing.py:
```diff
@@ -250,11 +250,11 @@ def create_final_segmentation_mask_with_boundaries(mask_image):
 
     edges = feature.canny(refined_mask[:, :, 0], sigma=3).astype(np.uint8)
-    contours, _ = cv2.findContours(edges, cv2.RETR_TREE, cv2.CHAIN_APPROX_NONE)
+    contours, _ = cv2.findContours(edges, cv2.RETR_TREE, cv2.CHAIN_APPROX_NONE)[-2:] # a more recent cv2 version has 3 returned values
```
Is there a specific reason for updating cv2 to a more recent version?

wendywwang replied:
@ricdodds no (it's simply that our environment uses a newer version, and downgrading it is more complicated than downgrading a regular package like pandas, maybe because of it os dependencies?); this change is compatible for both new and old cv2

ricdodds commented on deepliif/models/networks.py:
```diff
@@ -321,7 +328,7 @@ class ResnetGenerator(nn.Module):
     """
 
-    def __init__(self, input_nc, output_nc, ngf=64, norm_layer=nn.BatchNorm2d, use_dropout=False, n_blocks=6, padding_type='reflect'):
+    def __init__(self, input_nc, output_nc, ngf=64, norm_layer=nn.BatchNorm2d, use_dropout=False, n_blocks=6, padding_type='zero'):
```
We could leave here the original default and eventually pass the `padding_type` as a param of the CLI command?

```suggestion
    def __init__(self, input_nc, output_nc, ngf=64, norm_layer=nn.BatchNorm2d, use_dropout=False, n_blocks=6, padding_type='reflect'):
```

wendywwang replied:
@ricdodds I like this suggestion. It provides more flexibility and preserves the old default behavior! Can you make `padding_type` a training param? I will then pick up your change and add special handling for deterministic training (maybe warning messages emphasizing that the current padding type "reflect" does not have a deterministic implementation)

Parmida93 replied:
Please add padding_type as a parameter to training (in my offline version, I have done this). The 'reflect' works better than 'zero', only in case we need to train the deterministic version of the model, we will change the padding_type to 'zero'.

ricdodds reviewed (changes requested):
These changes look good to me although I haven't been able to test the changes on the visualizer yet. Just added some minor comments.

wendywwang commented on deepliif/util/__init__.py:
```diff
@@ -26,31 +26,6 @@ def chunker(iterable, size):
         yield iterable[i::size]
 
 
-def tensor2im(input_image, imtype=np.uint8):
```
@ricdodds 🤔 if you run `deepliif train` do you get the same error complaining about util.tensor2im cannot be found? I was testing `python cli.py train`, and not sure if it would lead to a slightly different behavior

wendywwang commented on run_visualizer_local.py:
```diff
@@ -0,0 +1,56 @@
+import os
```
@ricdodds if you passed `--remote True`, the training command will not start a visdom session and pass values over, so this script is used for this situation to start visdom then scan and consume updated pickle snapshots

wendywwang replied:
wrapped in a new command `deepliif visualize` (under review atm on our forked repo)

wendywwang commented on deepliif/data/__init__.py:
```diff
@@ -74,11 +77,33 @@ def __init__(self, dataset, batch_size, serial_batches, num_threads, max_dataset
+        # control randomness: https://pytorch.org/docs/stable/notes/randomness.html#dataloader
+        import numpy as np
+        import random
+        def seed_worker(worker_id):
```
@ricdodds I guess so.. this code snippet is copied directly from pytorch's doc (line 85)

wendywwang commented on deepliif/models/__init__.py:
```diff
@@ -122,7 +123,7 @@ def load_eager_models(model_dir, devices):
 
 @lru_cache
-def init_nets(model_dir, eager_mode=False):
+def init_nets(model_dir, eager_mode=True):
```
fixed

wendywwang commented:
the latest commit includes code changes for statistics:
1. FID: switched to the original implementation (https://github.com/bioinf-jku/TTUR/blob/master/fid.py) for better compatibility (scipy)
  - the implementation in `fid_official_tf.py` has dependency on `scipy.misc.imread` which is deprecated in scipy 1.3.0 (https://docs.scipy.org/doc/scipy/release.1.3.0.html#backwards-incompatible-changes, released in May 2019)
  - the original FID implementation added (`fid.py`) has moved to `imageio.imread`, the suggested alternative after scipy deprecated `imread`
2. inception score: added a try except to either import inception_v3 from keras (old) or from tensorflow.keras (new), compatible for both older and newer keras

wendywwang commented:
Hmmm we've been testing DDP directly using `cli.py`, because `torchrun` expects a training script.

Would it be fine if we suggest the users to clone the repo (instead of doing pip install) if they want to run DDP? Another possibility would be to wrap the execution into cli.py.. let me think it through. Please comment if you have any thoughts.

Parmida93 commented on deepliif/models/networks.py:
```diff
@@ -340,7 +347,7 @@ def __init__(self, input_nc, output_nc, ngf=64, norm_layer=nn.BatchNorm2d, use_d
             use_bias = norm_layer == nn.InstanceNorm2d
 
-        model = [nn.ReflectionPad2d(3),
+        model = [nn.ZeroPad2d(3) if os.environ['DEEPLIIF_SEED'] != 'None' else nn.ReflectionPad2d(3),
```
Please use the new version of the networks.py. I just updated the file and added padding_type to define_G and Resnet_Generator. If the padding_type is 'reflect' we use the ReflectionPad2d otherwise, we use ZeroPad2d. Instead of checking the DEEPLIIF_SEED, we can check the padding_type parameter. I believe it would make more sense and give more flexibility to the user, meaning that even without using the deterministic version, they can use the ZeroPad2d by only setting the padding_type to 'zero'. We can change the padding_type to 'zero' in the set_seed() function to make sure that the user will not receive any error by forgetting to set the padding_type to 'zero'.

Parmida93 commented on deepliif/models/base_model.py:
```diff
@@ -128,23 +136,51 @@ def get_current_losses(self):
                     torch.save(net.cpu().state_dict(), save_path)
+                    custom_save(save_path, self.remote_transfer_cmd_module, self.remote_transfer_cmd_function)
+
+
```
Did you check the save function? Does it work properly? If I'm not wrong, the other time, you mentioned that you have some issues with the save function.

wendywwang replied:
@Parmida93 
See this mentioned in the PR request:
> Implemented but not enabled features:
> ...
> * save models from only 1 process rather than all processes (hangs at the next iteration)
>   * note that saving from all processes vs. 1 doesn't make a difference if we simply want to have the models persisted, because all processes will end up with exactly the same parameters after gradient sync

Nothing goes wrong in terms of saving if you run our PR version of the code in the tests we did. However, we might miss some cases by chance. Please do bring it up if you see something weird in model saving using this code.

wendywwang commented:
Regarding the concern around torchrun here (https://github.com/nadeemlab/DeepLIIF/pull/10#issuecomment-1024500204):
I wrapped the execution of training in a new command `deepliif trainlaunch` that takes care of this, though may not look too pretty (under review atm on our forked repo)

ricdodds commented:
@wendywwang so sorry for the delay, is this ready for review?

wendywwang commented:
@ricdodds sorry I just came back from vacation :)

For now, everything except for the deterministic training config (--seed) should work, and please review when you have time.

Regarding deterministic training, basically we have this challenge caused by reflection pad2d. @Parmida93 I merged your change and tried it. The additional param `padding_type` is only effective in `deepliif/models/networks.py`. Can you or @ricdodds expose this parameter to the `train()` in `cli.py`, so that it's configurable on that level? 

Another minor comment on that change @Parmida93 is that currently although in the function description this `padding_type` can accept `replicate`, the logic doesn't take that into account. 
<img width="1005" alt="Screen Shot 2022-02-07 at 14 30 18" src="https://user-images.githubusercontent.com/97479603/152858375-53da1de2-7d46-46a3-82a8-e4e5d5a0aa27.png">

wendywwang commented:
### About `deepliif trainlaunch`
A new command `trainlaunch` is added that
1. parses the same set of options as `train` plus one more `--use-torchrun` for running DDP on local machine
2. executes `deepliif/train.py` (content equivalent to `train` in `cli.py`) through subprocess, with the parsed options appended (so `trainlaunch` effectively passes the received configurations over to `deepliif/train.py`)

In this way, no matter users prefer to run DP or DDP, the interface will be the same, both leveraging a deepliif command.

Note
- `trainlaunch` can be used for single process / DP too, and in this case `deepliif trainlaunch` is equivalent to `deepliif train`; once this implementation is approved, we can retire `deepliif train`, or replace it with `trainlaunch`.
- for deepliif **code developers**, `trainlaunch` has to be tested after you get `deepliif` built and installed; see the function comment section for `trainlaunch` in `cli.py`.

ricdodds commented:
@wendywwang let me know if this works
https://github.com/wendywwang/DeepLIIF/pull/28

ricdodds reviewed (approved):
Added a couple of small comments on how to potentially simplify the trainlaunch command but other than that all looks good. My only remaining concern has to do with duplicating the training code but at the same time we're thinking how to restructure the repo to better support devs workflows both for research and prod so we should be able to tackle this issue during that refactoring. Great work!

ricdodds commented:
This could be a potential solution for the duplicate code
https://click.palletsprojects.com/en/8.0.x/advanced/#forwarding-unknown-options

wendywwang commented on cli.py:
```diff
@@ -208,6 +275,158 @@ def train(dataroot, name, gpu_ids, checkpoints_dir, targets_no, input_nc, output
+                    k = d_map_key[k] if k in d_map_key.keys() else k
+                    options.append(f'--{k.replace("_","-")} {element}')
+
+    options = ' '.join(options)
```
@ricdodds let me try this! wasn't sure if this is compatible with click

wendywwang replied:
fixed; more complicated than that one liner, but still I think with sys we only need to specify the args to skip which is much simpler than before where I specified the name mapping and key value mapping for many args @ricdodds

wendywwang commented:
> This could be a potential solution for the `train.py` duplicated code https://click.palletsprojects.com/en/8.0.x/advanced/#forwarding-unknown-options

@ricdodds at the moment we have to use a train.py because the torchrun command expects a script to run.. I don't think click's unknown options could save us from that? 🤔

Parmida93 commented:
> @ricdodds sorry I just came back from vacation :)
> 
> For now, everything except for the deterministic training config (--seed) should work, and please review when you have time.
> 
> Regarding deterministic training, basically we have this challenge caused by reflection pad2d. @Parmida93 I merged your change and tried it. The additional param `padding_type` is only effective in `deepliif/models/networks.py`. Can you or @ricdodds expose this parameter to the `train()` in `cli.py`, so that it's configurable on that level?
> 
> Another minor comment on that change @Parmida93 is that currently although in the function description this `padding_type` can accept `replicate`, the logic doesn't take that into account. <img alt="Screen Shot 2022-02-07 at 14 30 18" width="1005" src="https://user-images.githubusercontent.com/97479603/152858375-53da1de2-7d46-46a3-82a8-e4e5d5a0aa27.png">

Thanks for mentioning it. I will remove the replicate from the choice options.

Parmida93 commented:
> > This could be a potential solution for the `train.py` duplicated code https://click.palletsprojects.com/en/8.0.x/advanced/#forwarding-unknown-options
> 
> @ricdodds at the moment we have to use a train.py because the torchrun command expects a script to run.. I don't think click's unknown options could save us from that? 🤔

We will definitely keep train.py, and we don't have any plan for removing it and we need that the torchrun works on the train.py script.

wendywwang commented on cli.py:
```diff
@@ -208,6 +275,158 @@ def train(dataroot, name, gpu_ids, checkpoints_dir, targets_no, input_nc, output
+    options = ' '.join(options)
+
+    #### locate train.py
+    import deepliif
```
@ricdodds this doesn't work in my environment.. it complains about deepliif not having train; to make the code look a little better I'm switching to
```
path_train_py = deepliif.__path__[0] + '/train.py'
```
the `__path__` method returns a list of locations of this module like the following:
<img width="513" alt="Screen Shot 2022-02-10 at 13 51 10" src="https://user-images.githubusercontent.com/97479603/153476437-71c05dd0-694d-4db5-9221-8119f5f80353.png">

wendywwang replied:
fixed

wendywwang commented:
I will merge the conflicts very soon for you to take a final look at the commits after you approved @ricdodds @Parmida93 

Another thing that caught my attention is, the current `inference` function runs init_net() on every image and every tile of that image, unless pytorch has special handling at the backend to avoid doing the same initialization all over again. Is this something you plan to improve? Also, how do you handle this in torchserve, if it's not done in entirely the same way?

@drewmibm @rich-nieto-ibm fyi

wendywwang commented:
@ricdodds let me try a rebase and see if github likes it

wendywwang commented:
closed as this is moved to #12
